R/deploymentTarget.R

```diff
@@ -1,93 +1,241 @@
 # calculate the deployment target based on the passed parameters and
 # any saved deployments that we have
-deploymentTarget <- function(recordPath = ".",
-                             appId = NULL,
-                             appName = NULL,
-                             appTitle = NULL,
-                             envVars = NULL,
-                             account = NULL,
-                             server = NULL,
-                             forceUpdate = FALSE,
-                             error_call = caller_env()) {
+deploymentTarget <- function(
+  recordPath = ".",
+  appId = NULL,
+  appName = NULL,
+  appTitle = NULL,
+  envVars = NULL,
+  account = NULL,
+  server = NULL,
+  forceUpdate = FALSE,
+  error_call = caller_env()
+) {
 
-  appDeployments <- deployments(
+  if (!is.null(appId)) {
+    return(deploymentTargetFromAppId(
+      recordPath = recordPath,
+      appId = appId,
+      appName = appName,
+      appTitle = appTitle,
+      envVars = envVars,
+      account = account,
+      server = server,
+      error_call = error_call
+    ))
+  }
+
+  if (!is.null(appName)) {
+    return(deploymentTargetFromAppName(
+      recordPath = recordPath,
+      appName = appName,
+      appTitle = appTitle,
+      envVars = envVars,
+      account = account,
+      server = server,
+      forceUpdate = forceUpdate,
+      error_call = error_call
+    ))
+  }
+
+  # No identifying appId or appName.
+
+  # When there are existing deployments, ask the user to select one and use
+  # it. Only deployments associated with locally configured account+server
+  # combinations are considered.
+  allDeployments <- deployments(
     appPath = recordPath,
-    nameFilter = appName,
     accountFilter = account,
     serverFilter = server
   )
+  if (nrow(allDeployments) > 0) {
+    deployment <- disambiguateDeployments(allDeployments, error_call = error_call)
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    accountDetails <- accountInfo(deployment$account, deployment$server)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
+  }
 
-  if (nrow(appDeployments) == 0) {
-    fullAccount <- findAccount(account, server)
-    if (is.null(appName)) {
-      appName <- defaultAppName(recordPath, fullAccount$server)
-    } else {
-      check_string(appName, call = error_call)
-    }
+  # Otherwise, identify a target account (given just one available or prompted
+  # by the user), generate a name, and locate the deployment.
+  accountDetails <- accountInfo(account, server)
+  appName <- generateAppName(appTitle, recordPath, accountDetails$name, unique = FALSE)
+  return(deploymentTargetFromAppName(
+    recordPath = recordPath,
+    appName = appName,
+    appTitle = appTitle,
+    envVars = envVars,
+    account = accountDetails$name,
+    server = accountDetails$server,
+    forceUpdate = forceUpdate,
+    error_call = error_call
+  ))
```

Comment:
```suggestion
  deploymentTargetFromAppName(
    recordPath = recordPath,
    appName = appName,
    appTitle = appTitle,
    envVars = envVars,
    account = accountDetails$name,
    server = accountDetails$server,
    forceUpdate = forceUpdate,
    error_call = error_call
  )
```

Reply:
Would you always favor removing the final return and prefer the implicit return? I like the fact that it's visually obvious that the two calls to `deploymentTargetFromAppName()` look the same -- with similar call signature and having their results immediately returned.

Reply:
Sorry, I misread this because I was looking at the unified diff.